omnigibson/utils/motion_planning_utils.py

```diff
@@ -56,44 +123,54 @@ def state_valid_fn(q):
     ss.setStateValidityChecker(ob.StateValidityCheckerFn(state_valid_fn))
 
     si = ss.getSpaceInformation()
+    si.setMotionValidator(CustomMotionValidator(si, space))
     planner = ompl_geo.RRTConnect(si)
     ss.setPlanner(planner)
 
-    start = ob.State(space)
-    start().setX(start_conf[0])
-    start().setY(start_conf[1])
-    start().setYaw(T.wrap_angle(start_conf[2]))
+    start = create_state(space, start_conf[0], start_conf[1], T.wrap_angle(start_conf[2]))
     print(start)
 
-    goal = ob.State(space)
-    goal().setX(end_conf[0])
-    goal().setY(end_conf[1])
-    goal().setYaw(T.wrap_angle(end_conf[2]))
+    goal = create_state(space, end_conf[0], end_conf[1], T.wrap_angle(end_conf[2]))
     print(goal)
 
     ss.setStartAndGoalStates(start, goal)
     if not state_valid_fn(start()) or not state_valid_fn(goal()):
         return
-
-    # this will automatically choose a default planner with
-    # default parameters
+      
     solved = ss.solve(planning_time)
 
     if solved:
         # try to shorten the path
         ss.simplifySolution()
-        # print the simplified path
         sol_path = ss.getSolutionPath()
         return_path = []
+
         for i in range(sol_path.getStateCount()):
             x = sol_path.getState(i).getX()
             y = sol_path.getState(i).getY()
             yaw = sol_path.getState(i).getYaw()
             return_path.append([x, y, yaw])
-        return remove_unnecessary_rotations(return_path)
+        return_path = remove_unnecessary_rotations(return_path)
+
+        # Validate return path
+        for i, pose in enumerate(return_path):
+            if i == 0:
+                final_ori =  CustomMotionValidator.get_angle_between_poses(pose, return_path[i+1])
+                if not CustomMotionValidator.is_valid_rotation(si, pose, final_ori):
+                    return None
+            elif i == len(return_path) - 1:
+                start_ori =  CustomMotionValidator.get_angle_between_poses(return_path[i-1], pose)
+                if not CustomMotionValidator.is_valid_rotation(si, [pose[0], pose[1], start_ori], pose[2]):
+                    return None
+            else:
+                start_ori =  CustomMotionValidator.get_angle_between_poses(return_path[i-1], pose)
+                final_ori =  CustomMotionValidator.get_angle_between_poses(pose, return_path[i+1])
+                if not CustomMotionValidator.is_valid_rotation(si, [pose[0], pose[1], start_ori], final_ori):
+                    return None
```

Comment:
This is not what I meant earlier - we should do our unnecessary rotation removal with this in mind (e.g. only remove rotations that don't make the path invalid). There is a deterministic way to do this that doesn't involve rejecting the path.

Reply:
My approach essentially involves doing this:

```
def remove_unnecessary_rotations(path):
  # Start at the same starting pose
  new_path = [path[0]]

  # Process every intermediate waypoint
  for i in range(1, len(path) - 1):
    # compute the yaw you'd be at when arriving into path[i] and departing from it
    arriving_yaw = CustomMotionValidator.get_angle_between_poses(path[i-1], path[i])
    departing_yaw = CustomMotionValidator.get_angle_between_poses(path[i], path[i+1])

    # check if you are able to make that rotation directly. 
    arriving_state = (path[i][0], path[i][1], arriving_yaw)
    if CustomMotionValidator.is_valid_rotation(si, arriving_state, departing_yaw):
      # Then use the arriving yaw directly
      new_path.append(arriving_state)
    else:
      # Otherwise, keep the waypoint
      new_path.append(path[i])

  # Don't forget to add back the same ending pose
  new_path.append(path[-1])

  return new_path
```

Reply:
If you do this, then you don't need to revalidate the path.